allennlp_models/generation/models/copynet_seq2seq.py

```diff
@@ -514,13 +524,33 @@ def _forward_loss(
             (batch_size, self._target_vocab_size), fill_value=1.0, dtype=torch.bool
         )
 
+        # Initialize target predictions with the start index.
+        # shape: (batch_size,)
+        last_predictions = source_mask.new_full(
+            (batch_size,), fill_value=self._start_index, dtype=torch.long
+        )
+
         step_log_likelihoods = []
         for timestep in range(num_decoding_steps):
-            # shape: (batch_size,)
-            input_choices = target_tokens["tokens"]["tokens"][:, timestep]
-            # Get mask tensor indicating which instances were copied.
-            # shape: (batch_size,)
-            copied = ((input_choices == self._oov_index) & (target_to_source.sum(-1) > 0)).long()
+            if (
+                self.training
+                and self._scheduled_sampling_ratio
+                and torch.rand(1).item() < self._scheduled_sampling_ratio
+            ):
+                # Use gold tokens at test time and at a rate of 1 - _scheduled_sampling_ratio
+                # during training.
+                # shape: (batch_size,)
+                input_choices = last_predictions
```

Comment:
Ah, I realized this implementation doesn't work, because `last_predictions` is never updated. I would have had to take the index of the token with the highest probability for this timestep under the model. Something like:

```python
last_predictions = torch.max(torch.cat((copy_scores, generation_scores), -1), -1)
```

@epwalsh does this make sense?

Reply:
Gotcha. Could I just return [`log_probs`](https://github.com/allenai/allennlp-models/blob/86444374c487f1d6991cd83b7a3da80b026effc2/allennlp_models/generation/models/copynet_seq2seq.py#L430) from `_get_ll_contrib()`? Its computed like: `log_probs = util.masked_log_softmax(all_scores, mask)`.

Reply:
Yes, good point.

Reply:
Awesome, just pushed that change.